### PR TITLE
PHPStan level 6

### PR DESCRIPTION
**Pull request by spawnia**

### CODE REVIEW

**Comment by spawnia** [comment]
@simPod I am having a hard time with some of the template/generics stuff, see `phpstan-baseline.neon`.

Having to specify a generic type for `NodeList` is tricky, because type hinting it as the most open `NodeList<Node>` causes errors when trying to pass in a more specific instance, e.g. `NodeList<VariableNode>`. I was able to solve a similar case in `ListOfType` by marking its generic parameter as `@template-covariant`, but that is not possible in `NodeList` due to it accepting writes. Do you have an idea how to solve this?

If you have any ideas on how to fix the other errors I put in the baseline, that would also be great. Some of them just seem like we are hitting the limits of PHPStan's capabilities?

**Comment by simPod** [comment]
I'm currently doing some winter sports, will check it out ASAP when I'm
back (few days)

**Comment by spawnia** [comment]
Thank you, have fun in the mountains 🏂

**Comment by simPod** [comment]
> due to it accepting writes. Do you have an idea how to solve this?

this is actually correct as `NodeList<VariableNode>` has the same contract as `NodeList<Node>`

1. With current state we could accept `mixed` (`template T`) and throw something similar like java does `java.lang.ArrayStoreException` if the node value is not of the same type as others nor subtype of `Node`
2. Also, we can consider dropping array access and making the NodeList immutable. I tried to search for usages and it seems we don't need `accepting writes`

File: src/Executor/ReferenceExecutor.php [hunk 1]
```diff
@@ -926,6 +927,7 @@ function ($previous, $value) use ($callback) {
     /**
      * Complete a list value by completing each item in the list with the inner type.
      *
+     * @param ListOfType<Type             &OutputType> $returnType
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Executor/ReferenceExecutor.php:930]: I will try https://github.com/webonyx/graphql-php/issues/1020 after this PR is merged.

File: src/Utils/ASTDefinitionBuilder.php [hunk 2]
```diff
@@ -95,60 +90,14 @@ public function buildDirective(DirectiveDefinitionNode $directiveNode): Directiv
...
-    /**
-     * Given an ast node, returns its string description.
-     */
-    private function getDescription(Node $node): ?string
-    {
-        if (isset($node->description)) {
-            return $node->description->value;
-        }
-
-        if (isset($this->options['commentDescriptions'])) {
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Utils/ASTDefinitionBuilder.php:115]: Following the reference implementation, I removed this deprecated option.

File: tests/Executor/Promise/SyncPromiseTest.php [hunk 3]
```diff
@@ -418,16 +433,16 @@ public function testPendingPromiseThen(): void
...
+        self::assertValidPromise($nextPromise, 1, null, SyncPromise::FULFILLED);
+        self::assertValidPromise($nextPromise2, 1, null, SyncPromise::FULFILLED);
+        self::assertValidPromise($nextPromise3, 2, null, SyncPromise::FULFILLED);
+        self::assertValidPromise($nextPromise4, 3, null, SyncPromise::FULFILLED);
     }
 
     public function testRunEmptyQueue(): void
     {
-        SyncPromise::runQueue();
+        $this->expectNotToPerformAssertions();
```
[the remark this answers is not in the document]
Comment (spawnia) [reply tests/Executor/Promise/SyncPromiseTest.php]: Reintroduced `assertDidNotCrash()`

**Comment by spawnia** [comment]
1. We should avoid runtime checks, they are bad for performance. I think it is acceptable to have code that *might* be misused, as long as we have the tests to ensure we don't.
2. We make heavy use of AST manipulation in `nuwave/lighthouse`, including appending to `NodeList` via `[]=`.

**Comment by simPod** [comment]
It we want to keep it this way we should maybe document the NodeList is not designed well as consumers will get the same errors as we do now from static analysis.

> The following example shows the disparity between the validation rules for array and generic classes

The example is about something else than we're dealing with though. And in out situations arrays are not relevant. The problem is LSP that requires method parameters to accept the same as parent does. So since `NodeList<Node>::add(Node $node)` accepts any Node, `NodeList<SubNode>::add(Node $node)` must accept it as well. This works the same with inheritance as generics.


> I think of the type safety afforded to us by PHPStan as a tool to help us write better, more robust code. If it fails to achieve that in some cases, we can choose to ignore it.

We can allow it because php does not care, at least yet. E.g. in scala this would no compile as it's against LSP.

**Comment by spawnia** [comment]
I have been pondering the comparison with arrays some more. In contrast to classes whose type parameters are set at construction, an array (and the current implementation of `NodeList`) allows dynamically changing it. When pushing a `FieldNode` to a `NodeList<ArgumentNode>`, it becomes a `NodeList<FieldNode|ArgumentNode>`. I don't think this is ever sensible, but it would work without crashing. Anyways, that nuance is irrelevant for our case, `NodeList` always contains a single type - and the type does not change at runtime.

I desire the following two properties from `NodeList`, and they seem to be at odds with each other:
1. Allow specifying the contained type in a structured way that is accessible to static analysis. Very useful when dealing the AST. I don't want to erase the genericness.
2. Allow code to accept a `NodeList` with any generic type as long as it is only read.

I dug around for ways to resolve 2., and found something that somewhat works. However, it surfaces what I believe to be a bug in PHPStan - see https://github.com/phpstan/phpstan/issues/6237. We can ignore that in the baseline for now though, we will have a robust solution in place for when it is fixed.

File: src/Language/AST/Node.php [hunk 4]
```diff
@@ -92,36 +93,29 @@ protected static function cloneValue($value)
...
-        $tmp = (array) $this;
-
-        if ($this->loc !== null) {
-            $tmp['loc'] = [
-                'start' => $this->loc->start,
-                'end'   => $this->loc->end,
-            ];
-        }
+        return $this->toArray();
+    }
```
Comment (simPod) [inline src/Language/AST/Node.php:105]: Is this actually needed?

**Review by simPod — APPROVED** [review]
This is already too huge so we can 🚢 

_About the NodeList, I wonder if I'm right (generic NodeList abuses LSP and should not compile in decent language) or you're right (phpstan needs some feature in order to define that)_ 😜

File: CHANGELOG.md [hunk 5]
```diff
@@ -42,6 +45,7 @@ You can find and compare releases at the [GitHub release page](https://github.co
 - Add SDL validation rule `UniqueOperationTypes` (#995)
 - Add ability to remove custom validation rules after adding them via `DocumentValidator::removeRule()`
 - Allow lazy enum values
+- Make `Node` implement `JsonSerializable`
```
Comment (simPod) [inline CHANGELOG.md:48]: can we specify why eventually? If I was to chose I'd not add the interface

**Comment by spawnia** [comment]
I think the definition of `NodeList` is fine, it is a mutable container around a generic type. Thus, its template type is neither covariant nor covariant. PHPStan lacks the ability for a method to express that it uses `NodeList` exclusively as if its template type were covariant. Java actually has such a feature, which is why I opened https://github.com/phpstan/phpstan/issues/6240.

File: src/Language/AST/Node.php (hunk 4 above)
Comment (spawnia) [reply src/Language/AST/Node.php:105]: Yeah, it makes the JSON representation more compact. In contrast to the default serialization, null values and some runtime only values (such as pointers to previous and next tokens in `Location`) are excluded.

File: CHANGELOG.md (hunk 5 above)
Comment (spawnia) [reply CHANGELOG.md:48]: Added a docblock to explain why it is useful

**Pull request merged**